### PR TITLE
Update the Language Server API

### LINKED ISSUE #3304: Discrepancies in PR 3286
After the first round of QA/Acceptance Review testing on PR #3286 (https://www.pivotaltracker.com/story/show/180916527), focusing on the `library/getPackage` message, I found some discrepancies between what is described in the design doc, what happens actually, and what is described in the API docs. They fall mostly in two general categories:

 1. The yaml example from Design Doc does not work verbatim; various tweaks seem needed, and I'm not sure why and whether they are documented anywhere;
 2. The format of the response from `library/getPackage` seems to not match the format described in the API doc, both in field names and in field types (at various levels of nesting). 

A list of example cases I observed comes below.

### ***API documentation of `ComponentGroups.newGroups` field name is inconsistent with `.new` field being apparently returned instead***
```
{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"new":[{"module":"Gruuup 1","exports":["foo"]}]},"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"exports":["foo"]}]}}}}
```

### ***Yaml semantics different than in Design Doc***
Snippet similar as presented in Design Doc does not work:
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
        exports:
          - foo

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":null,"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":{"exports":["foo"]}}]}}}}
```
Instead, snippet with different indentation (which conveys different semantics in YAML, as can be seen comparing the `"raw"` field in the response below vs. above) works:
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
      exports:
        - foo

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"new":[{"module":"Gruuup 1","exports":["foo"]}]},"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"exports":["foo"]}]}}}}
```
Similarly, adding a `shortcut` as in design doc does not work:
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
      exports:
        - foo:
            shortcut: f

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":null,"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"exports":[{"foo":{"shortcut":"f"}}]}]}}}}
```
but instead semantically different YAML document seems to work:
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
      exports:
        - foo:
          shortcut: f

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"new":[{"module":"Gruuup 1","exports":[{"name":"foo","shortcut":"f"}]}]},"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"exports":[{"foo":null,"shortcut":"f"}]}]}}}}
```

### ***Nonconforming response returned***
In response, at path `result.componentGroups.new[0].exports[0]` note a string value `"foo"`, where expected a `Component` containing `{"name":"foo"}`.
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
      exports:
        - foo

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"new":[{"module":"Gruuup 1","exports":["foo"]}]},"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"exports":["foo"]}]}}}}
```
In response, at path `result.componentGroups.new[0].exports[1]` note a string value `"bar"`, where expected a `Component` containing `{"name":"bar"}`.
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
      exports:
        - foo:
          shortcut: f
        - bar

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"new":[{"module":"Gruuup 1","exports":[{"name":"foo","shortcut":"f"},"bar"]}]},"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"exports":[{"foo":null,"shortcut":"f"},"bar"]}]}}}}
```

### ***Empty response returned with no error message, even though component groups at least partially correct***
Is it described somewhere how we want to behave in case of partially correct input? Do we want to somehow provide error message to user to guide them what they did wrong? Notably, the `extends:` section in the document below was copied from the design doc.
```
name: Footest
component-groups:
  new:
    - Gruuup 1:
      color: '#aabbcc'
      icon: some-icon
      exports:
        - foo:
          shortcut: f
        - bar
    - Le gruppe 2:
      exports:
        - foo
  extends:
    - Base.Group 2:
      exports:
        - foo

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":null,"raw":{"name":"Footest","component-groups":{"new":[{"Gruuup 1":null,"color":"#aabbcc","icon":"some-icon","exports":[{"foo":null,"shortcut":"f"},"bar"]},{"Le gruppe 2":null,"exports":["foo"]}],"extends":[{"Base.Group 2":null,"exports":["foo"]}]}}}}
```

### ***API docs for `extendedGroups` field don't match actual semantics***
see https://github.com/enso-org/enso/blob/abbb3a467915eabf37e91329317ba6c3274fc1e0/docs/language-server/protocol-language-server.md#componentgroups
In example below:
 1. Instead of `.extendedGroup`, field named `extends` is returned
 2. At `result.componentGroups.extends[0].module`, expected a `ModuleReference` object according to API docs, but got a string `"Standard.Base.Main"` instead
```
name: Footest
component-groups:
  extends:
    - Standard.Base.Main:
      exports:
      - name: foo

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"extends":[{"module":"Standard.Base.Main","exports":["foo"]}]},"raw":{"name":"Footest","component-groups":{"extends":[{"Standard.Base.Main":null,"exports":[{"name":"foo"}]}]}}}}
```
 

### ***Error code should be 8007, is 1000, for LocalLibraryNotFound***
Per definition of [LocalLibraryNotFound](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md#locallibrarynotfound) message, referenced in the [errors section of `library/getPackage` message](https://github.com/enso-org/enso/blob/abbb3a467915eabf37e91329317ba6c3274fc1e0/docs/language-server/protocol-language-server.md#errors-74). Got code 1000 in transcript below:
```
{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Fooblah","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"error":{"code":1000,"message":"Local library [local.Fooblah] has not been found."}}
```

### ***Returned `componentGroups` field's empty value is sometimes `{}`, sometimes `null`***
Not strictly error, but possibly surprising behavior.
```
{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{},"raw":{"name":"Footest"}}}

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{},"raw":{"name":"Footest","component-groups":null}}}

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{},"raw":{"name":"Footest","component-groups":{"new":null}}}}

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":null,"raw":{"name":"Footest","component-groups":{"new":["Gruuup 1"]}}}}
```

### PR DESCRIPTION
**Pull request by 4e6**
[ci no changelog needed]
### Pull Request Description

close #3304

Changelog:
- Fix project.yaml codecs
- Simplify the JSONRPC API
- Handle the `LocalLibraryNotFound` errors and return appropriate JSONRPC response
- Fix typo in `Standard.Searcher` library 

### Checklist

Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), and [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [ ] If GUI codebase was changed: Enso GUI was tested when built using BOTH `./run dist` and `./run watch`.

### CODE REVIEW

File: docs/language-server/protocol-language-server.md [hunk 1]
```diff
@@ -1427,79 +1424,55 @@ interface NamedEdition {
...
-The definition of a single component group.
+The component group provided by a library.
 
 ```typescript
-interface ComponentGroup {
-  /** The module name containing the declared componennts. */
+interface LibraryComponentGroup {
+  /**
+   * Thf fully qualified module name. A string consisting of a namespace and
+   * a lirary name separated by the dot <namespace>.<library name>,
```
Comment (radeusgd) [inline docs/language-server/protocol-language-server.md]:
```suggestion
   * a library name separated by the dot <namespace>.<library name>,
```

**Comment by akavel** [comment]
Most of the discrepancies seem resolved now ❤️ 

However, I still observe a few unexpected behaviors (also the improvements helped me discover one more case that I did not try before):


### [unfixed] _**Error code should be 8007, is 1000, for [LocalLibraryNotFound](https://github.com/enso-org/enso/blob/develop/docs/language-server/protocol-language-server.md#locallibrarynotfound)**_

(This appears to still not be fixed - testing is done on commit 3bd47c290.)

Per definition of [LocalLibraryNotFound](https://github.com/enso-org/enso/blob/3bd47c290e74e6ed99050f916a5a40c2efae769c/docs/language-server/protocol-language-server.md#locallibrarynotfound) message, referenced in the errors section of `library/getPackage` message, expected code 8007, but got code 1000 in transcript below:

```
{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Fooblah","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"error":{"code":1000,"message":"Local library [local.Fooblah] has not been found."}}
```


### ***Bad error code when wrong edition supplied***

According to API docs, expected [EditionNotFoundError](https://github.com/enso-org/enso/blob/3bd47c290e74e6ed99050f916a5a40c2efae769c/docs/language-server/protocol-language-server.md#editionnotfounderror), which is described as having code 8001, but got code 1000:
```
{"jsonrpc":"2.0","id":4,"method":"editions/listDefinedComponents","params":{"edition":{"type":"NamedEdition","editionName":"asdf"}}}
{"jsonrpc":"2.0","id":4,"error":{"code":1000,"message":"The edition was not found."}}
```

### ***Non-null component group returned when empty `exports:` added***

Not 100% sure if that's bad or not, but seems somewhat surprising: adding an empty `exports:` section in YAML leads to `componentGroups` being returned with non-`null` contents. Removing this line results in `null` contents. In both cases, there are no "LibraryComponent" items defined by the user in the `package.yaml` file, so it's not clear why the response would be different.
```
name: Footest
component-groups:
  new:
    - Review 2:
        exports:

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":{"newGroups":[{"library":"local.Footest","module":"Review 2"}]},"raw":{"name":"Footest","component-groups":{"new":[{"Review 2":{"exports":null}}]}}}}

name: Footest
component-groups:
  new:
    - Review 2:

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":null,"raw":{"name":"Footest","component-groups":{"new":[{"Review 2":null}]}}}}

name: Footest
component-groups:
  new:
    - Review 2

{"jsonrpc":"2.0","id":5,"method":"library/getPackage","params":{"namespace":"local","name":"Footest","version":{"type":"LocalLibraryVersion"}}}
{"jsonrpc":"2.0","id":5,"result":{"license":null,"componentGroups":null,"raw":{"name":"Footest","component-groups":{"new":["Review 2"]}}}}
```

**Comment by 4e6** [comment]
@akavel Thank you for testing that! I fixed the error messages, but I think that the behavior with empty exports is ok and we should allow it

File: engine/language-server/src/main/scala/org/enso/languageserver/libraries/LocalLibraryManager.scala [hunk 2]
```diff
@@ -202,12 +202,16 @@ class LocalLibraryManager(
...
     } yield {
-      if (config.componentGroups.isLeft) {
-        logger.error(
-          s"Failed to parse library [$libraryName] component groups."
-        )
+      config.componentGroups match {
+        case Left(error) =>
+          logger.error(
+            s"Failed to parse library [$libraryName] component groups " +
+            s"(reason: ${error.message})"
```
Comment (radeusgd) [inline engine/language-server/src/main/scala/org/enso/languageserver/libraries/LocalLibraryManager.scala]: Missing dot at end of sentence.

**Review by radeusgd — APPROVED** [review]
Had just a quick look given that Mateusz already checked it, but it looks good.

**Comment by akavel** [comment]
@4e6 The error https://github.com/enso-org/enso/blob/a9f38056c3d08bcdf395e667ad87a11fe9ee71d2/docs/language-server/protocol-language-server.md#editionnotfounderror specifies a field named `"payload"`, but got a field named `"data"` in the response:

```
{"jsonrpc":"2.0","id":4,"method":"editions/listDefinedComponents","params":{"edition":{"type":"NamedEdition","editionName":"asdf"}}}
{"jsonrpc":"2.0","id":4,"error":{"data":{"editionName":"asdf"},"code":8001,"message":"Edition [asdf] could not be found."}}
```

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/enso-org/enso">enso/protocol-language-server.md at a9f38056c3d08bcdf395e667ad87a11fe9ee71d2 · enso-org/enso</a></strong></div><div>Hybrid visual and textual functional programming. Contribute to enso-org/enso development by creating an account on GitHub.</div></blockquote>

**Comment by 4e6** [comment]
This is actually a general discrepancy and true for all returned errors (they all have `data` field instead of `payload`). I will fix that in a separate PR because it will touch unrelated code (probably just the tests, but still)

**Pull request merged**